File: flitetrakr_sync/src/test/java/com/github/albertosh/flitetrakr/sync/usecases/connectionsbelowprice/ConnectionsBelowPriceUseCaseTest.java
package com.github.albertosh.flitetrakr.sync.usecases.connectionsbelowprice;

import com.github.albertosh.flitetrakr.model.MultipleConnections;
import com.github.albertosh.flitetrakr.sync.services.IConnectionService;
import com.github.albertosh.flitetrakr.sync.usecases.MockService;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.CheapestConnectionUseCaseError;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.CheapestConnectionUseCaseOutput;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.ICheapestConnectionUseCase;

import org.junit.Before;
import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ConnectionsBelowPriceUseCaseTest {

    private IConnectionService service;
    private ICheapestConnectionUseCase cheapestConnectionUseCase;
    private ConnectionsBelowPriceUseCase connectionsBelowPriceUseCase;

    @Before
    public void setUp() throws Exception {
        service = MockService.doMock();
        cheapestConnectionUseCase = mock(ICheapestConnectionUseCase.class);
        connectionsBelowPriceUseCase = new ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service);
    }

    @Test
    public void existingConnections() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "NUE";
        final String TO = "LHR";
        final Integer PRICE = 170;

        // Force that at least a cheap connection exists
        when(cheapestConnectionUseCase.execute(any()))
                .thenReturn(new CheapestConnectionUseCaseOutput.Builder()
                        .price(0)
                        .cities(new ArrayList<>())
                        .build());

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(2));

        assertThat(output.get(0).getCities(), contains("NUE", "FRA", "LHR"));
        assertThat(output.get(0).getPrice(), is(70));

        assertThat(output.get(1).getCities(), contains("NUE", "FRA", "LHR", "NUE", "FRA", "LHR"));
        assertThat(output.get(1).getPrice(), is(163));
    }

    @Test
    public void whenTheCheapestConnectionIsOverMinPriceTheAlgorithmHalts() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "NUE";
        final String TO = "LHR";
        final Integer PRICE = 170;

        // Force that the cheapest connection is more expensive than the min price
        when(cheapestConnectionUseCase.execute(any()))
                .thenReturn(new CheapestConnectionUseCaseOutput.Builder()
                        .price(PRICE + 10)
                        .cities(new ArrayList<>())
                        .build());

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(0));
    }

    @Test
    public void aNonExistingConnectionsThrowsAnError() throws Exception, CheapestConnectionUseCaseError {
        final String FROM = "NUE";
        final String TO = "DBX";
        final Integer PRICE = 1000;

        // Force that there is no connection
        when(cheapestConnectionUseCase.execute(any()))
                .thenThrow(CheapestConnectionUseCaseError.connectionNotFound());

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();

        ConnectionsBelowPriceUseCaseError error = null;
        try {
            connectionsBelowPriceUseCase.execute(input);
        } catch (ConnectionsBelowPriceUseCaseError connectionsBelowPriceUseCaseError) {
            error = connectionsBelowPriceUseCaseError;
        }

        assertThat(error, is(not(nullValue())));
        assertThat(error.getErrorType(), is(ConnectionsBelowPriceUseCaseError.ErrorType.CONNECTION_NOT_FOUND));
    }

    @Test
    public void sameFromAndTo() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "LHR";
        final String TO = "LHR";
        final Integer PRICE = 100;

        when(cheapestConnectionUseCase.execute(any()))
                .thenReturn(new CheapestConnectionUseCaseOutput.Builder()
                        .price(0)
                        .cities(new ArrayList<>())
                        .build());

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(2));

        assertThat(output.get(0).getCities(), contains("LHR"));
        assertThat(output.get(0).getPrice(), is(0));

        assertThat(output.get(1).getCities(), contains("LHR", "NUE", "FRA", "LHR"));
        assertThat(output.get(1).getPrice(), is(93));
    }
}

File: flitetrakr_sync_kotlin/src/test/kotlin/com/github/albertosh/flitetrakr/usecases/ConnectionsBelowPriceUseCaseTest.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.model.MultipleConnections
import com.github.albertosh.flitetrakr.services.IConnectionService
import com.nhaarman.mockito_kotlin.any
import com.nhaarman.mockito_kotlin.doThrow
import com.nhaarman.mockito_kotlin.mock
import io.kotlintest.mock.`when`
import io.kotlintest.specs.ShouldSpec
import org.mockito.Mockito.verifyZeroInteractions

class ConnectionsBelowPriceUseCaseTest : ShouldSpec() {

    init {
        "ConnectionBelowPrice" {
            should("throw an error if a connection doesn't exist") {
                val service: IConnectionService = mockService()
                val cheapestConnectionUseCase = mock<ICheapestConnectionUseCase>()
                val useCase = ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service)

                // Force that there is no connection
                doThrow(
                        CheapestConnectionUseCaseError.connectionNotFound
                ).`when`(cheapestConnectionUseCase).execute(any())

                val exception: ConnectionsBelowPriceUseCaseError? =
                        try {
                            useCase.execute(
                                    ConnectionsBelowPriceUseCaseInput(
                                            from = "NUE",
                                            to = "DBX",
                                            price = Int.MAX_VALUE
                                    ))
                            null
                        } catch (e: ConnectionsBelowPriceUseCaseError) {
                            e
                        }

                exception shouldBe ConnectionsBelowPriceUseCaseError.connectionNotFound
            }

            should("look for connections below some price if at least a connection below that price exists") {
                val service: IConnectionService = mockService()
                val cheapestConnectionUseCase = mock<ICheapestConnectionUseCase>()
                val useCase = ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service)

                // Force that at least a cheap connection exists
                `when`(
                        cheapestConnectionUseCase.execute(any()))
                        .thenReturn(CheapestConnectionUseCaseOutput(listOf(), 0))

                val output = useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "NUE",
                                to = "LHR",
                                price = 170))

                output shouldBe ConnectionsBelowPriceUseCaseOutput(listOf(
                        MultipleConnections(
                                cities = listOf("NUE", "FRA", "LHR"),
                                price = 70
                        ),
                        MultipleConnections(
                                cities = listOf("NUE", "FRA", "LHR", "NUE", "FRA", "LHR"),
                                price = 163
                        )
                ))
            }

            should("halt when the cheapest connection is over maximum price") {
                val service: IConnectionService = mockService()
                val cheapestConnectionUseCase = mock<ICheapestConnectionUseCase>()
                val useCase = ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service)

                `when`(
                        cheapestConnectionUseCase.execute(any()))
                        .thenReturn(CheapestConnectionUseCaseOutput(listOf(), 200))

                val output = useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "NUE",
                                to = "LHR",
                                price = 170))

                output shouldBe ConnectionsBelowPriceUseCaseOutput(emptyList())
                verifyZeroInteractions(service)
            }

            should("return the existing connections ordered below some price even if the destination and origin is the same") {
                val service: IConnectionService = mockService()
                val cheapestConnectionUseCase = mock<ICheapestConnectionUseCase>()
                val useCase = ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service)

                // Force that at least a cheap connection exists
                `when`(
                        cheapestConnectionUseCase.execute(any()))
                        .thenReturn(CheapestConnectionUseCaseOutput(listOf(), 0))

                val output = useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "LHR",
                                to = "LHR",
                                price = 100))

                output shouldBe ConnectionsBelowPriceUseCaseOutput(listOf(
                        MultipleConnections(
                                cities = listOf("LHR"),
                                price = 0
                        ),
                        MultipleConnections(
                                cities = listOf("LHR", "NUE", "FRA", "LHR"),
                                price = 93
                        )
                ))
            }

        }
    }
}
